### ISSUE TITLE
Potential Security Gap

### ISSUE THREAD

**Issue by LTsLlama** [MicrosoftDX/AuthBot#26]
Hello,
I might be wrong here, but after doing more testing on magic number auth flow I found a potential security gap.
@matvelloso I'm adding you to this discussion as you wrote this logic and would be the best source of truth.

# Auth Flow
Here is a summary of the current auth flow how I understand it:
1. On a message to the bot, the bot checks if a token is available in the bot user data for the specific user id.
2. If no token is available, a auth link is built and posted to the user. 
**IMPORTANT:** The auth link contains all the data relative to the conversation including the user id and therefore the magic number is required to ensure no one highjacks the conversation.

>  Assume a hacker intercepts the request and now have the user's conversation data.

3. The user clicks the auth link opening a browser session where they enter their user detail and the AD token and conversation data is redirected to the bot (not the chat client.)
4. When the bot receives the token and conversation data it performs two tasks:
a. The token is saved against the user id in the bot data store marked as validated = false and a magic number is provided on the web request.
b. The conversation data is used to find the existing conversation and send a message to the user's chat client asking for the magic number.

> At this point the hacker can talk to the bot as the user, but is rejected because the token has not been validated yet and is therefore rejected. The hacker do not have the magic number and therefore cannot validate the token. All good thus far. 

5.  The user copies the magic number and sends it to the bot on the same conversation.
6. The bot receives the magic number from the user, checks that it's the same and marks the token saved against the user id as validated. 
7. The user can now talk to the bot and use the token saved in the bot user data against their user id.

# Problem / Assumption
The token saved against a specific user id is now validated and can be used by anyone who talks to the bot with the same user id.

> The hacker still has the user id and can now use the validated token to do malicious shit.

# In summary
The magic number flow ensures the token first has to be validated by the user who made the auth request, but once the token has been validated a security gap is opened.

This gap could be closed by validating the magic number on each message, instead of just once. Not sure how this would be implemented as each channel will have to persist and send the magic number on each message.

Is my understanding of the flow correct? Is there a gap or am I missing something?

I found this issue by using the DirectLine channel. 

Cheers,

Douw

**Comment by matvelloso** [comment]
First of all, many thanks for raising this. It's great that people are thinking about the potential gaps in the security. 

Once the magic number is validated, in order for a hacker to inject anything into that conversation with that stolen user ID, channel ID and conversation ID, the hacker would have to either:

1-Make a call to the bot framework pretending to the the channel. This would be impossible in all channels, including the direct line, as the hacker won't have an authentication token that the real user has. So any attempt to send that message would be rejected by the bot framework

2-Make a call to your bot's web service, pretending to be bot framework. This call will be rejected because the hacker doesn't have the application ID/security key that bot framework has (your web service will reject it unless it hasn't been properly configured according to the bot framework guidance)

So yes, the fact that the hacker is holding a user id, conversation id and channel id is very far from ideal and a thing we would like to address in the future. Still, I don't see how that will allow them to work around the magic number... Did I miss anything? Do you have a repro where you can do that?

Thanks again

**Comment by LTsLlama** [comment]
Agree, the direct line secret / token adds another safety layer.
The channel Id is public knowledge and the conversation Id is only needed before token validation, not after, so the user id is the only real unknown.

I forget to mention that we are working on a multi-tenant bot where the direct line channel is shared between lots of users and the hacker could easily have access to the direct line bot.
Even on a single tenant bot we don't want users being able to impersonate another user, by just having their user id.

This gap is not an issue on any of the channels that has their own authentication layer. I.e. You can only use the Skype user id if you are already logged into Skype.

@matvelloso I suppose my real question is, how does the magic number flow improve security?
The magic number ensures the conversation cannot be high jacked, but the attacker can just start a new conversation and get access to the users token.

PS. We found this issue by downloading the standard [Bot Framework WebChat](https://github.com/Microsoft/BotFramework-WebChat). The custom direct line WebChat use a static user Id if not specified as parameter and therefore as soon as one developer was signed in everyone shared his token (luckily a fake demo user). The Bot Framework WebChat channel auto generate a random user id eliminating this issue.

**Comment by matvelloso** [comment]
I see. In this case I'd recommend persisting the authentication token per conversation, not per user. That way you should be protected from someone attempting to perform the attack you mentioned. Having said, that, typically we don't recommend AuthBot for directLine clients as they tend to already have their own authentication and can instead just hand over the existing token to the bot, giving a full SSO experience.

On your question about the need for the magic number, this thread discusses some scenarios where this helps protecting the user: https://github.com/MicrosoftDX/AuthBot/issues/23

**Comment by LTsLlama** [comment]
Thank you. I think it be worthwhile listing these limitations in the read me.

**Comment by LTsLlama** [comment]
@matvelloso I closed this gap by saving the conversation reference (ResumtionCookie) on Azure Blob storage instead of adding it to the auth request. This ensures none of the conversation / user data is exposed on the auth request anymore.
For DirectLine I'm encrypting the user id on a cookie to ensure it cannot be accessed, so I believe even DirectLine is now secure.
My version is here: https://github.com/LiveTiles/AuthBot
 
Additionally:
1. Updated ResumptionCookie (obsolete) to ConversationReference.
2. Removed hack of posting to user when an empty message is received from the OAuthCallbackController, instead posting to the user directly on the conversation.

I cannot create a pull request for the improvements as I introduced an Azure Storage dependency.

**Comment by matvelloso** [comment]
That's a good approach and we discussed about making this the default, but the concern with the added complexity made me fear most developers would end up hacking an alternative, less secure path instead. 

I'm thinking it's still worth adding some of these directions in the readme page. 

Thanks again
